ide/parsing.indexing/src/org/netbeans/modules/parsing/spi/indexing/IndexabilityQueryImplementation.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.parsing.spi.indexing;
+
+import java.net.URL;
+import javax.swing.event.ChangeListener;
+import org.openide.filesystems.FileObject;
+
+/**
+ * Determine whether files should be skipped for indexing.
+ * <p>
+ * Global lookup is used to find all instances of
+ * IndexabilityQueryImplementation.
+ * </p>
+ * <p>
+ * Threading note: implementors should avoid acquiring locks that might be held
+ * by other threads. Generally treat this interface similarly to SPIs in
+ * {@link org.openide.filesystems} with respect to threading semantics.
+ * </p>
+ *
+ * @see org.netbeans.api.queries.IndexabilityQuery
+ * @since org.netbeans.modules.parsing.indexing 9.22.0
+ */
+public interface IndexabilityQueryImplementation {
+
+    /**
+     * Determine if the supplied file object should be indexed.
+     * 
+     * @param fo file object to be indexed
+     * @return true if indexing should be prevented
+     */
+    default boolean preventIndexing(FileObject fo) {
```

Comment:
Is the "duality" of two `default` methods (which [are usually bad](http://wiki.apidesign.org/wiki/DefaultMethods) in an API as they increase fuzziness) necessary? Since [ability to return non-existing files](https://bits.netbeans.org/12.5/javadoc/org-openide-filesystems/apichanges.html#getFileObjectNonExisting) we may not need the `URI` variant at all.